File: Level Designer/src/Sprite.java
// ======================================= //
// Melvin the Marvellous Monster from Mars //
// Level Designer                          //
//                                         //
// Author: Kevin Scroggins                 //
// E-Mail: nitro404@hotmail.com            //
// Date: April 11, 2010                    //
// ======================================= //

import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import javax.imageio.ImageIO;

/**
 * The Sprite class represents a 2D image which can be drawn onto the screen.
 * 
 * @author Kevin Scroggins
 */
public class Sprite {
	
	// local variables
	private File imageFile;
	private BufferedImage image;
	private String name;
	private String parentName;
	private int index;
	private int type;
	
	/** Unknown Sprite type constant. */
	public static int TYPE_UNKNOWN = -1;
	/** SpriteSheet image Sprite type constant. */
	public static int TYPE_SHEET = 0;
	/** Tiled Sprite type constant. */
	public static int TYPE_TILE = 1;
	/** Object Sprite type constant. */
	public static int TYPE_OBJECT = 2;
	/** Player Sprite type constant. */
	public static int TYPE_PLAYER = 3;
	/** Disguise Sprite type constant. */
	public static int TYPE_DISGUISE = 4;
	/** Pet Sprite type constant. */
	public static int TYPE_PET = 5;
	/** AI Sprite type constant. */
	public static int TYPE_AI = 6;
	/** Item Sprite type constant. */
	public static int TYPE_ITEM = 7;
	
	
	/**
	 * Constructs a new Sprite from a specified file and directory.
	 * 
	 * @param fileName the sprite image file name.
	 * @param spriteDirectory the directory to read the file from.
	 */
	public Sprite(String fileName, String spriteDirectory) {
		this(fileName, spriteDirectory, -1);
	}
	
	/**
	 * Constructs a Sprite from a specified file and directory and assigns a specified index to it.
	 * 
	 * @param fileName the sprite image file name.
	 * @param spriteDirectory the directory to read the file from.
	 * @param index the index of the sprite within its parent SpriteSheet.
	 */
	public Sprite(String fileName, String spriteDirectory, int index) {
		this.name = null;
		this.parentName = null;
		this.index = index;
		this.type = TYPE_UNKNOWN;
		
		// generate the file path
		String imagePath = spriteDirectory;
		if(imagePath.charAt(imagePath.length() - 1) != '\\' && imagePath.charAt(imagePath.length() - 1) != '/') {
			imagePath += '\\';
		}
		imagePath += fileName;
		
		// create a file pointing to the image
		this.imageFile = new File(imagePath);
		
		// validate the file
		if(!this.imageFile.exists() || !this.imageFile.isFile()) {
			System.out.println("ERROR: Sprite: \"" + this.imageFile.getPath() + "\" does not exist or is an invalid image.");
			System.exit(1);
		}
		
		// read the image and store it
		try {
			this.image = ImageIO.read(this.imageFile);
		}
		catch(IOException e) {
			System.out.println("ERROR: Unable to read sprite image: \"" + this.imageFile.getPath() + "\".");
			System.exit(1);
		}
	}
	
	/**
	 * Constructs a Sprite from a BufferedImage.
	 * 
	 * @param image the image to assign to the Sprite.
	 */
	public Sprite(BufferedImage image) {
		this(image, -1);
	}
	
	/**
	 * Constructs a Sprite from a BufferedImage and assigns a specific index to it.
	 * 
	 * @param image the image to assign to the Sprite.
	 * @param index the index of the Sprite within its parent SpriteSheet.
	 */
	public Sprite(BufferedImage image, int index) {
		this.name = null;
		this.parentName = null;
		this.index = index;
		this.type = TYPE_UNKNOWN;
		
		if(image == null) {
			System.out.println("ERROR: Sprite cannot be created from null image.");
			System.exit(1);
		}
		
		this.image = image;
	}
	
	/**
	 * Constructs a Sprite from a sub-image of an image and assigns a specific index to it.
	 * 
	 * @param image the image to assign to the Sprite.
	 * @param index the index of the Sprite within its parent SpriteSheet.
	 * @param x the x offset of the sub-image.
	 * @param y the y offset of the sub-image.
	 * @param width the width of the sub-image.
	 * @param height the height of the sub-image.
	 */
	public Sprite(BufferedImage image, int index, int x, int y, int width, int height) {
		this.name = null;
		this.parentName = null;
		this.index = index;
		this.type = TYPE_UNKNOWN;
		
		if(image == null) {
			System.out.println("ERROR: Sprite cannot be created from null image.");
			System.exit(1);
		}
		
		this.image = image.getSubimage(x, y, width, height);
	}
	
	/**
	 * Returns the width of the Sprite.
	 * 
	 * @return the width of the Sprite.
	 */
	public int getWidth() {
		return this.image.getWidth();
	}
	
	/**
	 * Returns the height of the Sprite.
	 * 
	 * @return the height of the Sprite.
	 */
	public int getHeight() {
		return this.image.getHeight();
	}
	
	/**
	 * Returns the size of the Sprite.
	 * 
	 * @return the size of the Sprite.
	 */
	public Dimension getDimensions() {
		return new Dimension(this.image.getWidth(), this.image.getHeight());
	}
	
	/**
	 * Checks to see if the Sprite is tiled or not.
	 * 
	 * @return true if the Sprite is tiled.
	 */
	public boolean isTiled() {
		return this.type == TYPE_TILE;
	}
	
	/**
	 * Returns the name of the Sprite.
	 * 
	 * @return the name of the Sprite.
	 */
	public String getName() {
		return this.name;
	}
	
	/**
	 * Returns the name of the SpriteSheet the Sprite belongs to.
	 * 
	 * @return the name of the SpriteSheet the Sprite belongs to.
	 */
	public String getParentName() {
		return this.parentName;
	}
	
	/**
	 * Returns the index of the Sprite within the SpriteSheet it belongs to.
	 * 
	 * @return the index of the Sprite within the SpriteSheet it belongs to.
	 */
	public int getIndex() {
		return this.index;
	}
	
	/**
	 * Returns the Sprite type.
	 * 
	 * @return the Sprite type.
	 */
	public int getType() {
		return this.type;
	}
	
	/**
	 * Changes the name assigned to the current Sprite.
	 * 
	 * @param name the name to assign to the Sprite.
	 */
	public void setName(String name) {
		this.name = name;
	}
	
	/**
	 * Changes the parent name assigned to the current Sprite.
	 * 
	 * @param parentName the parent name to assign to the Sprite.
	 */
	public void setParentName(String parentName) {
		this.parentName = parentName;
	}
	
	/**
	 * Changes the indexed location of the Sprite within it's parent SpriteSheet.
	 * 
	 * @param index the location of the Sprite within its parent SpriteSheet.
	 */
	public void setIndex(int index) {
		this.index = index;
	}
	
	/**
	 * Changes the sprite type.
	 * 
	 * @param type the type to change the Sprite to.
	 */
	public void setType(int type) {
		if(type == TYPE_SHEET ||
		   type == TYPE_TILE ||
		   type == TYPE_OBJECT ||
		   type == TYPE_PLAYER ||
		   type == TYPE_DISGUISE ||
		   type == TYPE_PET ||
		   type == TYPE_AI ||
		   type == TYPE_ITEM) {
			this.type = type;
		}
		else {
			this.type = TYPE_UNKNOWN;
		}
	}
	
	/**
	 * Get the image stored inside of the current Sprite.
	 * 
	 * @return the image representing the current Sprite. 
	 */
	public BufferedImage getImage() {
		return this.image;
	}
	
	/**
	 * Converts a String type into an integer representing a constant value of the same Sprite type.
	 * 
	 * @param data the String representation of the Sprite type.
	 * @return the integer representation of the Sprite Type.
	 */
	public static int parseType(String data) {
		if(data == null || data.trim().length() == 0) { return TYPE_UNKNOWN; }
		String typeString = data.trim();
		
		if(typeString.equalsIgnoreCase("SpriteSheet")) {
			return TYPE_SHEET;
		}
		else if(typeString.equalsIgnoreCase("Tile")) {
			return TYPE_TILE;
		}
		else if(typeString.equalsIgnoreCase("Object")) {
			return TYPE_OBJECT;
		}
		else if(typeString.equalsIgnoreCase("Player")) {
			return TYPE_PLAYER;
		}
		else if(typeString.equalsIgnoreCase("Disguise")) {
			return TYPE_DISGUISE;
		}
		else if(typeString.equalsIgnoreCase("Pet")) {
			return TYPE_PET;
		}
		else if(typeString.equalsIgnoreCase("AI")) {
			return TYPE_AI;
		}
		else if(typeString.equalsIgnoreCase("Item")) {
			return TYPE_ITEM;
		}
		else {
			return TYPE_UNKNOWN;
		}
	}
	
	/**
	 * Renders the Sprite onto the specified Graphics component.
	 * 
	 * @param g the Graphics component to render to.
	 * @param x the x coordinate to draw the Sprite at.
	 * @param y the y coordinate to draw the Sprite at.
	 */
	public void paintOn(Graphics g, int x, int y) {
		((Graphics2D) g).drawImage(image, null, x, y);
	}
	
	/**
	 * Renders the Sprite onto the specified Graphics component, centred around the specified position.
	 * 
	 * @param g the Graphics component to render to.
	 * @param x the x coordinate to draw the Sprite at.
	 * @param y the y coordinate to draw the Sprite at.
	 */
	public void paintOnCentered(Graphics g, int x, int y) {
		((Graphics2D) g).drawImage(image, null, (int) (x - (image.getWidth() / 2.0)), (int) (y - (image.getHeight() / 2.0)));
	}
	
}

File: Game Engine/Source/Sprite.cpp
// ======================================= //
// Melvin the Marvellous Monster from Mars //
//                                         //
// Author: Kevin Scroggins                 //
// E-Mail: nitro404@hotmail.com            //
// Date: April 11, 2010                    //
// ======================================= //

#include "Sprite.h"

// sprite type constants
int Sprite::TYPE_UNKNOWN = -1;
int Sprite::TYPE_SHEET = 0;
int Sprite::TYPE_TILE = 1;
int Sprite::TYPE_OBJECT = 2;
int Sprite::TYPE_PLAYER = 3;
int Sprite::TYPE_DISGUISE = 4;
int Sprite::TYPE_PET = 5;
int Sprite::TYPE_AI = 6;
int Sprite::TYPE_ITEM = 6;

// constructor for creating a sprite from a given file name and sprite directory
Sprite::Sprite(const char * fileName,
			   const char * directory,
			   LPDIRECT3DDEVICE9 d3dDevice)
				: texture(NULL),
				  sprite(NULL),
				  name(NULL),
				  parentName(NULL),
				  index(-1),
				  type(TYPE_UNKNOWN),
				  xOffset(-1),
				  yOffset(-1),
				  width(0),
				  height(0),
				  ownsSprite(true) {
	if(fileName != NULL) {
		D3DXCreateSprite(d3dDevice, &sprite);

		string spriteFileName;
		if(directory != NULL) {
			spriteFileName.append(directory);
			if(_stricmp(spriteFileName.substr(spriteFileName.length() - 1, spriteFileName.length()).c_str(), ("\\")) != 0 &&
			   _stricmp(spriteFileName.substr(spriteFileName.length() - 1, spriteFileName.length()).c_str(), ("/")) != 0) {
				spriteFileName.append("\\");
			}
		}
		spriteFileName.append(fileName);


		if(FAILED(D3DXCreateTextureFromFile(d3dDevice, spriteFileName.c_str(), &texture))) {
			quit("Error", "Error loading sprite texture \"%s\".", spriteFileName.c_str());
		}

		D3DSURFACE_DESC description;
		texture->GetLevelDesc(0, &description);
		this->width = description.Width;
		this->height = description.Height;
	}
	
	D3DXMatrixScaling(&flipMatrix, -1, 1, 1);
}

// constructor for creating a subsprite from an already existing sprite's directx texture and sprite
Sprite::Sprite(int x,
			   int y,
			   int spriteWidth,
			   int spriteHeight,
			   LPDIRECT3DTEXTURE9 externalTexture,
			   LPD3DXSPRITE externalSprite)
				: xOffset(x),
				  yOffset(y),
				  name(NULL),
				  parentName(NULL),
				  index(-1),
				  type(TYPE_UNKNOWN),
				  width(spriteWidth),
				  height(spriteHeight),
				  texture(externalTexture),
				  sprite(externalSprite),
				  ownsSprite(false) { D3DXMatrixScaling(&flipMatrix, -1, 1, 1); }

// destructor
Sprite::~Sprite() {
	if(ownsSprite) {
		if(texture != NULL) { texture->Release(); }
		if(sprite != NULL) { sprite->Release(); }
	}
	if(name != NULL) { delete [] name; }
	if(parentName != NULL) { delete [] parentName; }
}

// get the sprite's x offset (width / 2)
int Sprite::getOffsetX() const { return xOffset; }

// get the sprite's y offset (height / 2)
int Sprite::getOffsetY() const { return yOffset; }

// get the sprite's width
int Sprite::getWidth() const { return width; }

// get the sprite's height
int Sprite::getHeight() const { return height; }

// check if the sprite is tiled
bool Sprite::isTiled() { return type == TYPE_TILE; }

// get the sprite's name
char * Sprite::getName() const { return this->name; }

// get the name of the sprite sheet which the sprite belongs to
char * Sprite::getParentName() const { return this->parentName; }

// get the index of the sprite within its parent sprite sheet
int Sprite::getIndex() const { return this->index; }

// get the type of the sprite
int Sprite::getType() const { return this->type; }

// set the name of the sprite
void Sprite::setName(char * name) {
	if(name == NULL) { return; }
	if(this->name != NULL) {
		delete [] this->name;
	}
	this->name = new char[strlen(name) + 1];
	strcpy_s(this->name, strlen(name) + 1, name);
}

// set the name of the sprite sheet which the sprite belongs to
void Sprite::setParentName(char * parentName) {
	if(parentName == NULL) {
		this->parentName = NULL;
		return;
	}
	if(this->parentName != NULL) {
		delete [] this->parentName;
	}
	this->parentName = new char[strlen(parentName) + 1];
	strcpy_s(this->parentName, strlen(parentName) + 1, parentName);
}

// set the index of the sprite in its parent sprite sheet
void Sprite::setIndex(int index) {
	this->index = (index < -1) ? -1 : index;
}

// set the type of the sprite
void Sprite::setType(int type) {
	if(type == TYPE_SHEET ||
	   type == TYPE_TILE ||
	   type == TYPE_OBJECT ||
	   type == TYPE_PLAYER ||
	   type == TYPE_DISGUISE ||
	   type == TYPE_PET ||
	   type == TYPE_AI ||
	   type == TYPE_ITEM) {
		this->type = type;
	}
	else {
		this->type = TYPE_UNKNOWN;
	}
}

// get the sprite's directx texture handle
LPDIRECT3DTEXTURE9 Sprite::getTexture() const { return texture; }

// get the sprite's directx sprite handle
LPD3DXSPRITE Sprite::getSprite() const { return sprite; }

// parse the sprite's type from a string and return it
int Sprite::parseType(const char * data) {
	int type = TYPE_UNKNOWN;
	if(data == NULL) { return type; }
	char * typeString = strtrimcpy(data);
	
	if(_stricmp(typeString, "SpriteSheet") == 0) {
		type = TYPE_SHEET;
	}
	else if(_stricmp(typeString, "Tile") == 0) {
		type = TYPE_TILE;
	}
	else if(_stricmp(typeString, "Object") == 0) {
		type = TYPE_OBJECT;
	}
	else if(_stricmp(typeString, "Player") == 0) {
		type = TYPE_PLAYER;
	}
	else if(_stricmp(typeString, "Disguise") == 0) {
		type = TYPE_DISGUISE;
	}
	else if(_stricmp(typeString, "Pet") == 0) {
		type = TYPE_PET;
	}
	else if(_stricmp(typeString, "AI") == 0) {
		type = TYPE_AI;
	}
	else if(_stricmp(typeString, "Item") == 0) {
		type = TYPE_ITEM;
	}
	delete [] typeString;
	return type;
}

// render the sprite using a pre-computed matrix
void Sprite::draw(D3DXMATRIX & transformationMatrix, LPDIRECT3DDEVICE9 d3dDevice) {
	if(texture != NULL) {
		// apply the transformation to the sprite
		sprite->SetTransform(&transformationMatrix);

		sprite->Begin(D3DXSPRITE_ALPHABLEND);

		// if the sprite is in its natural state, simply render it
		if(xOffset < 0 && yOffset < 0) {
			sprite->Draw(texture, NULL, NULL, NULL, D3DCOLOR_RGBA(255, 255, 255, 255));
		}
		// otherwise render the sprite within a bounding box (a sub-sprite)
		else {
			RECT spriteRect = {xOffset, yOffset, xOffset + width, yOffset + width};
			sprite->Draw(texture, &spriteRect, NULL, NULL, D3DCOLOR_RGBA(255, 255, 255, 255));
		}

		sprite->End();
	}
}

// render the sprite using a pre-defined collection of transformation attributes
void Sprite::draw(D3DXVECTOR2 * scale, D3DXVECTOR2 * scalingOffset, float rotationDegrees, D3DXVECTOR2 * rotationOffset, D3DXVECTOR2 * position, LPDIRECT3DDEVICE9 d3dDevice) {
	if(texture != NULL) {
		// compute the transformation to be applied to the sprite
		D3DXMatrixTransformation2D(&transformationMatrix, scalingOffset, 0, scale, rotationOffset, D3DXToRadian(rotationDegrees), position);

		// apply the transformation to the sprite
		sprite->SetTransform(&transformationMatrix);

		sprite->Begin(D3DXSPRITE_ALPHABLEND);

		// if the sprite is in its natural state, simply render it
		if(xOffset < 0 && yOffset < 0) {
			sprite->Draw(texture, NULL, NULL, NULL, D3DCOLOR_RGBA(255, 255, 255, 255));
		}
		// otherwise render the sprite within a bounding box (a sub-sprite)
		else {
			RECT spriteRect = {xOffset, yOffset, xOffset + width, yOffset + height};
			sprite->Draw(texture, &spriteRect, NULL, NULL, D3DCOLOR_RGBA(255, 255, 255, 255));
		}

		sprite->End();
	}
}

// render the sprite backwards using a pre-defined set of transformation attributes
void Sprite::drawBackwards(D3DXVECTOR2 * scale, D3DXVECTOR2 * scalingOffset, float rotationDegrees, D3DXVECTOR2 * rotationOffset, D3DXVECTOR2 * position, LPDIRECT3DDEVICE9 d3dDevice) {
	if(texture != NULL) {
		// compute the transformation to be applied to the sprite and then flip the transformation along the x axis
		D3DXVECTOR2 flippedPosition(position->x + (scalingOffset->x * 2), position->y);
		D3DXMatrixTransformation2D(&transformationMatrix, scalingOffset, 0, scale, rotationOffset, D3DXToRadian(rotationDegrees), &flippedPosition);
		D3DXMatrixMultiply(&flippedTransformationMatrix, &flipMatrix, &transformationMatrix);

		// apply the transformation to the sprite
		sprite->SetTransform(&flippedTransformationMatrix);

		sprite->Begin(D3DXSPRITE_ALPHABLEND);

		// if the sprite is in its natural state, simply render it
		if(xOffset < 0 && yOffset < 0) {
			sprite->Draw(texture, NULL, NULL, NULL, D3DCOLOR_RGBA(255, 255, 255, 255));
		}
		// otherwise render the sprite within a bounding box (a sub-sprite)
		else {
			RECT spriteRect = {xOffset, yOffset, xOffset + width, yOffset + height};
			sprite->Draw(texture, &spriteRect, NULL, NULL, D3DCOLOR_RGBA(255, 255, 255, 255));
		}

		sprite->End();
	}
}
